src/refda/exec.c

```diff
@@ -373,93 +375,434 @@ void *refda_exec_worker(void *arg)
-                sem_post(&(agent->rptgs_sem));
+    return 0;
+}
 
-                break;
+/**
+ * Schedule execution of a time based rule
+ */
+static int refda_exec_schedule_tbr(refda_agent_t *agent, refda_amm_tbr_desc_t *tbr, bool starting)
+{
+    // Do not schedule TBR if it has reached its execution threshold
+    if (refda_amm_tbr_desc_reached_max_exec_count(tbr))
+    {
+        CACE_LOG_INFO("TBR %p reached maximum execution count", tbr);
+        refda_exec_tbr_disable(agent, tbr);
+        return 0;
+    }
+
+    struct timespec schedtime;
+    int             result = refda_exec_tbr_next_scheduled_time(&schedtime, tbr, starting);
+    if (!result)
+    {
+        refda_timeline_event_t event = { .purpose      = REFDA_TIMELINE_TBR,
+                                         .ts           = schedtime,
+                                         .tbr.agent    = agent,
+                                         .tbr.tbr      = tbr,
+                                         .tbr.callback = refda_exec_tbr };
+        refda_timeline_push(agent->exec_timeline, event);
+    }
+
+    return result;
+}
+
+int refda_exec_tbr_enable(refda_agent_t *agent, refda_amm_tbr_desc_t *tbr)
+{
+    CHKERR1(tbr);
+    if (tbr->action.is_ref || tbr->action.as_lit.ari_type != CACE_ARI_TYPE_AC)
+    {
+        CACE_LOG_ERR("Invalid TBR %p action, unable to enable the rule", tbr);
+        return 1;
+    }
+
+    // Adjust rule state
+    tbr->enabled    = true;
+    tbr->exec_count = 0; // Ensure count is reset when rule is enabled
+    atomic_fetch_add(&agent->instr.num_tbrs, 1);
+
+    // Schedule initial rule execution
+    int result = refda_exec_schedule_tbr(agent, tbr, true);
+    return result;
+}
+
+int refda_exec_tbr_disable(refda_agent_t *agent, refda_amm_tbr_desc_t *tbr)
+{
+    CHKERR1(tbr);
+    tbr->enabled = false;
+    atomic_fetch_sub(&agent->instr.num_tbrs, 1);
+    return 0;
+}
+
+static int refda_exec_schedule_sbr(refda_agent_t *agent, refda_amm_sbr_desc_t *sbr);
+
+static int refda_exec_check_sbr_condition(refda_agent_t *agent, refda_amm_sbr_desc_t *sbr, cace_ari_t *result)
+{
+    refda_runctx_t runctx;
+    refda_runctx_init(&runctx);
+
+    if (refda_runctx_from(&runctx, agent, NULL))
+    {
+        return 2;
+    }
+
+    cace_ari_t ari_res = CACE_ARI_INIT_UNDEFINED;
+    int        res     = refda_eval_target(&runctx, &ari_res, &(sbr->condition));
+
+    if (res)
+    {
+        CACE_LOG_ERR("Unable to evaluate SBR condition");
+    }
+    else
+    {
+        const cace_amm_type_t *typeobj = cace_amm_type_get_builtin(CACE_ARI_TYPE_BOOL);
+        res                            = cace_amm_type_convert(typeobj, result, &ari_res);
+        if (res)
+        {
+            CACE_LOG_ERR("Unable to convert SBR condition result to boolean");
+        }
+    }
+
+    cace_ari_deinit(&ari_res);
+    refda_runctx_deinit(&runctx);
+
+    return res;
+}
+
+/** Begin a single execution of a state based rule
```

Comment:
For consistency in terminology, if this is periodically evaluating the SBR condition then this should not be called "execution" or named "exec". If the AMM doesn't provide specific guidance this could be called "check" or "evaluate" on the rule instance.